---
title: "Differential expression analysis using DEseq2"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library("ggplot2")
library("tximport")
library("DESeq2")
library("biomaRt")
library("data.table")
```

```{r}
mart <- biomaRt::useMart(
  biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "mmusculus_gene_ensembl",
  host="https://nov2020.archive.ensembl.org" # mm10
)


t2g <- biomaRt::getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id",
                                     "external_gene_name"), mart = mart)
t2g <- dplyr::rename(t2g, target_id = ensembl_transcript_id,
                     ens_gene = ensembl_gene_id, ext_gene = external_gene_name)
tx2gene <- t2g[1:3]
tx2gene <- subset(t2g, select = -c(ext_gene)) ## use ensemble gene ids for gene expr quantification
```

Create dataframe which will provide genomic locations of genes (used to annotated gene selection).
```{r}
geneloc <- function(){
  #mart <- biomaRt::useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
  mart <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl", mirror = "useast",host="https://nov2020.archive.ensembl.org")
  gloc <- biomaRt::getBM(attributes = c("ensembl_gene_id","external_gene_name", "chromosome_name",
                                         "start_position", "end_position","strand"), mart = mart)
  gloc <- dplyr::rename(gloc, ens_gene = ensembl_gene_id, ext_gene = external_gene_name,
                        chrom = chromosome_name, chromStart = start_position, chromEnd = end_position, strand=strand)
  return(gloc)
}
gloc <- geneloc()
rownames(gloc) <- gloc$ens_gene
```


```{r}
kallisto_dir = '/home/andrew/biology/xdf/rotation3/data/eye_organoids/RNAseq/Kallisto_Output'

days_reps <- c("0","0","0","1","1","1","2","2","2","3","3","3","4neg","4neg","4neg","4pos","4pos","4pos","5neg","5neg","5neg","5pos","5pos","5pos")
samples = cbind(list.files(kallisto_dir),days_reps)
colnames(samples) <- c('sample','condition')
samples <- as.data.frame(samples)

files <- file.path(kallisto_dir, samples$sample, "abundance.tsv")

names(files) <- paste0("sample", 1:24)
all(file.exists(files))

txi <- tximport(files, type = "kallisto", tx2gene = tx2gene,
                ignoreTxVersion = TRUE)
all.equal(txi$counts, txi$counts)
results_table <- txi$abundance



genes <- unique(t2g[, 2:3])
genes <- genes[order(genes$ens_gene),]
names(results_table) <- genes

## load into DEseq object ##
ddsTxi <- DESeqDataSetFromTximport(txi, colData = samples, design = ~ condition)
dds <- DESeq(ddsTxi, betaPrior=FALSE)
```

Generate normalized counts (this is not used by DEseq2 for differential expression analysis, but useful for 
visualization and further analysis).
```{r}
dds <- estimateSizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)
```

```{r}
## differential testing to identify important genes ##
#for eye field genes
DEseqD34p_unshrunk <- results(dds, contrast = c("condition", "3", "4pos"))
DEseqD44p_unshrunk <- results(dds, contrast = c("condition", "4neg", "4pos"))
DEseqD55p_unshrunk <- results(dds, contrast = c("condition", "5neg", "5pos"))

DEseqD34p <- lfcShrink(dds, contrast = c("condition", "3", "4pos"), res=DEseqD34p_unshrunk, type = "ashr")
DEseqD44p <- lfcShrink(dds, contrast = c("condition", "4neg", "4pos"), res=DEseqD44p_unshrunk, type = "ashr")
DEseqD55p <- lfcShrink(dds, contrast = c("condition", "5neg", "5pos"), res=DEseqD55p_unshrunk, type = "ashr")

fdr_thr <- 0.001
fdr_thr_down <- 0.001
fc_thr <-  1.5
fc_thr_down <-  1.5
bool_efup_selection <- (DEseqD34p$padj   < fdr_thr) & (DEseqD34p$log2FoldChange < -fc_thr) &
                       (DEseqD44p$padj   < fdr_thr) & (DEseqD44p$log2FoldChange < -fc_thr) &
                       (DEseqD55p$padj   < fdr_thr) & (DEseqD55p$log2FoldChange < -fc_thr)
bool_efdown_selection <- (DEseqD34p$padj < fdr_thr_down) & (DEseqD34p$log2FoldChange > fc_thr_down) &
                         (DEseqD44p$padj < fdr_thr_down) & (DEseqD44p$log2FoldChange > fc_thr_down) &
                         (DEseqD55p$padj < fdr_thr_down) & (DEseqD55p$log2FoldChange > fc_thr_down)

bool_efup_selection[is.na(bool_efup_selection)] <- FALSE
bool_efdown_selection[is.na(bool_efdown_selection)] <- FALSE
sum(bool_efup_selection)
sum(bool_efdown_selection)

deseq_efup_genes   <- rownames(DEseqD34p)[bool_efup_selection]
deseq_efdown_genes <- rownames(DEseqD34p)[bool_efdown_selection]

sort(gloc[deseq_efup_genes,]$ext_gene)
sort(gloc[deseq_efdown_genes,]$ext_gene)
```

Now determine a set of house-keeping genes (stable expression across all timepoints), to use as controls in downstream analyses.
```{r}
DEseqD01_unshrunk  <- results(dds, contrast = c("condition", "0", "1"))
DEseqD12_unshrunk  <- results(dds, contrast = c("condition", "1", "2"))
DEseqD23_unshrunk  <- results(dds, contrast = c("condition", "2", "3"))
DEseqD34p_unshrunk <- results(dds, contrast = c("condition", "3", "4pos"))
DEseqD34n_unshrunk <- results(dds, contrast = c("condition", "3", "4neg"))
DEseqD4n4p_unshrunk <- results(dds, contrast = c("condition", "4neg", "4pos"))
DEseqD5n5p_unshrunk <- results(dds, contrast = c("condition", "5neg", "5pos"))
DEseqD4n5n_unshrunk <- results(dds, contrast = c("condition", "4neg", "5neg"))
DEseqD4p5p_unshrunk <- results(dds, contrast = c("condition", "4pos", "5pos"))

DEseqD01  <- lfcShrink(dds, contrast = c("condition", "0", "1"), res=DEseqD01_unshrunk, type = "ashr")
DEseqD12  <- lfcShrink(dds, contrast = c("condition", "1", "2"), res=DEseqD12_unshrunk, type = "ashr")
DEseqD23  <- lfcShrink(dds, contrast = c("condition", "2", "3"), res=DEseqD23_unshrunk, type = "ashr")
DEseqD34p <- lfcShrink(dds, contrast = c("condition", "3", "4pos"), res=DEseqD34p_unshrunk, type = "ashr")
DEseqD34n <- lfcShrink(dds, contrast = c("condition", "3", "4neg"), res=DEseqD34n_unshrunk, type = "ashr")
DEseqD4n4p <- lfcShrink(dds, contrast = c("condition", "4neg", "4pos"), res=DEseqD4n4p_unshrunk, type = "ashr")
DEseqD5n5p <- lfcShrink(dds, contrast = c("condition", "5neg", "5pos"), res=DEseqD5n5p_unshrunk, type = "ashr")
DEseqD4n5n <- lfcShrink(dds, contrast = c("condition", "4neg", "5neg"), res=DEseqD4n5n_unshrunk, type = "ashr")
DEseqD4p5p <- lfcShrink(dds, contrast = c("condition", "4pos", "5pos"), res=DEseqD4p5p_unshrunk, type = "ashr")

hk_fc_thr <- 0.1
hk_expr_thr <- 30
bool_hk_selection <- (abs(DEseqD01$log2FoldChange) < hk_fc_thr) & (abs(DEseqD12$log2FoldChange) < hk_fc_thr) &
                     (abs(DEseqD23$log2FoldChange) < hk_fc_thr) & 
                     (abs(DEseqD34p$log2FoldChange) < hk_fc_thr) & (abs(DEseqD34n$log2FoldChange) < hk_fc_thr) &
                     (abs(DEseqD4n4p$log2FoldChange) < hk_fc_thr) & (abs(DEseqD5n5p$log2FoldChange) < hk_fc_thr) &
                     (abs(DEseqD4n5n$log2FoldChange) < hk_fc_thr) & (abs(DEseqD4p5p$log2FoldChange) < hk_fc_thr) &
                     (rowMeans(normalized_counts) > hk_expr_thr)

deseq_hk_genes   <- rownames(DEseqD34p)[bool_hk_selection]
sort(gloc[deseq_hk_genes,]$ext_gene)
```

Now determine set of genes that are differentially expressed on successive days. 
The transition day2-to-day3 may be important in priming for eye-field state.
```{r}
DEseqD01_unshrunk  <- results(dds, contrast = c("condition", "0", "1"))
DEseqD12_unshrunk  <- results(dds, contrast = c("condition", "1", "2"))
DEseqD23_unshrunk  <- results(dds, contrast = c("condition", "2", "3"))

DEseqD01  <- lfcShrink(dds, contrast = c("condition", "0", "1"), res=DEseqD01_unshrunk, type = "ashr")
DEseqD12  <- lfcShrink(dds, contrast = c("condition", "1", "2"), res=DEseqD12_unshrunk, type = "ashr")
DEseqD23  <- lfcShrink(dds, contrast = c("condition", "2", "3"), res=DEseqD23_unshrunk, type = "ashr")

bool_01_up_selection   <- (DEseqD01$padj   < fdr_thr) & (DEseqD01$log2FoldChange < -fc_thr)
bool_01_down_selection <- (DEseqD01$padj   < fdr_thr) & (DEseqD01$log2FoldChange >  fc_thr)

bool_12_up_selection   <- (DEseqD12$padj   < fdr_thr) & (DEseqD12$log2FoldChange < -fc_thr)
bool_12_down_selection <- (DEseqD12$padj   < fdr_thr) & (DEseqD12$log2FoldChange >  fc_thr)

bool_23_up_selection   <- (DEseqD23$padj   < fdr_thr) & (DEseqD23$log2FoldChange < -fc_thr)
bool_23_down_selection <- (DEseqD23$padj   < fdr_thr) & (DEseqD23$log2FoldChange >  fc_thr)

deseq_01_up_genes   <- rownames(DEseqD01)[bool_01_up_selection]
deseq_01_down_genes <- rownames(DEseqD01)[bool_01_down_selection]

deseq_12_up_genes   <- rownames(DEseqD12)[bool_12_up_selection]
deseq_12_down_genes <- rownames(DEseqD12)[bool_12_down_selection]

deseq_23_up_genes   <- rownames(DEseqD23)[bool_23_up_selection]
deseq_23_down_genes <- rownames(DEseqD23)[bool_23_down_selection]

sort(gloc[deseq_23_up_genes,]$ext_gene)
sort(gloc[deseq_23_down_genes,]$ext_gene)
```




```{r}
## set up gene names boolean factors that might be useful in downstream analyses
genes_data <- rownames(normalized_counts)

bool_efup_genes <- genes_data %in% deseq_efup_genes
bool_efdown_genes <- genes_data %in% deseq_efdown_genes
bool_hk_genes <- genes_data %in% deseq_hk_genes
bool_01_up_genes <- genes_data %in% deseq_01_up_genes
bool_01_down_genes <- genes_data %in% deseq_01_down_genes
bool_12_up_genes <- genes_data %in% deseq_12_up_genes
bool_12_down_genes <- genes_data %in% deseq_12_down_genes
bool_23_up_genes <- genes_data %in% deseq_23_up_genes
bool_23_down_genes <- genes_data %in% deseq_23_down_genes
```



## Write out full results to file for ease of downstream analysis

For potential TAD modelling, we record the TAD-id for each gene. This uses TAD boundaries 
derived by the Aerts Lab obtained from [here](https://github.com/aertslab/mucistarget/tree/master/data/tads), 
using data from [Bonev et al, 2017](https://dx.doi.org/10.1016%2Fj.cell.2017.09.043).
Note we have filled-in genomic intervals between TADs, as 'inter-TADs'.
```{r}
library(IRanges)
library(stringr)
library(GenomicRanges)
# tad_file <- '/home/andrew/biology/xdf/rotation3/mammalian_eftfs/notebooks/TADs-interTADs_mESC.Bonev_2017.bed'
tad_file <- '../data/TADs-interTADs_mESC.Bonev_2017.bed'
tmp_tads <- read.table(tad_file,header=FALSE)
colnames(tmp_tads) <- c('chrom','chromStart','chromEnd','source')
tmp_tads$chrom <- str_remove_all(tmp_tads$chrom, "chr")
tmp_gloc <- gloc[genes_data,c("ext_gene","ens_gene","chrom","chromStart","chromEnd")]

ranges_genes <- with(tmp_gloc,GRanges(chrom,IRanges(chromStart,chromEnd),ext_gene=ext_gene,ens_gene=ens_gene))
ranges_tads  <- with(tmp_tads,GRanges(chrom,IRanges(chromStart,chromEnd),source=source))

gene_tad_ov <- findOverlaps(ranges_genes, ranges_tads, type="within", select = "all")
gene_tad_ov <- as.matrix(gene_tad_ov)

tad_id    <- vector(length = length(genes_data))
for (i in seq(1:length(genes_data))) {
  if (i %in% gene_tad_ov[,1]) {
    tad_id[i] <- gene_tad_ov[,2][gene_tad_ov[,1] == i]
  }
  else {
    tad_id[i] <- -1
  }
}

tad_df <- as.data.frame(tad_id)
colnames(tad_df) <- c("tad-id-bonev")
```

Save normalized expression data, saving also genomic locations as well as indicators as to whether genes are differentially
expressed across timecourse or across GFP positive/negative lineages.
```{r eval=FALSE}
tmp_df <- normalized_counts
colnames(tmp_df) <- c("Day0_1","Day0_2","Day0_3",
                      "Day1_1","Day1_2","Day1_3",
                      "Day2_1","Day2_2","Day2_3",
                      "Day3_1","Day3_2","Day3_3",
                      "Day4_GFPneg_1","Day4_GFPneg_2","Day4_GFPneg_3",
                      "Day4_GFPpos_1","Day4_GFPpos_2","Day4_GFPpos_3",
                      "Day5_GFPneg_1","Day5_GFPneg_2","Day5_GFPneg_3",
                      "Day5_GFPpos_1","Day5_GFPpos_2","Day5_GFPpos_3")
tmp_genomic_locations <- gloc[genes_data,c("ext_gene","ens_gene","chrom","chromStart","chromEnd","strand")]

tmp_bool_df <- cbind(bool_efup_genes, bool_efdown_genes, 
                     bool_01_up_genes, bool_01_down_genes, 
                     bool_12_up_genes, bool_12_down_genes,
                     bool_23_up_genes, bool_23_down_genes,
                     bool_hk_genes)
colnames(tmp_bool_df) <- c("EF-up","EF-down",
                           "D0-D1-up","D0-D1-down",
                           "D1-D2-up","D1-D2-down",
                           "D2-D3-up","D2-D3-down",
                           "HK")

tmp_df <- cbind(tmp_genomic_locations, tmp_df)
tmp_df <- cbind(tmp_df,tmp_bool_df)

tmp_df <- cbind(tmp_df, tad_df)

filename <- "../data/RNAseq-DEseq2-preprocessed-norm-counts.tsv"
write.table(tmp_df,filename,quote=FALSE,sep='\t',col.names = NA)
rm(tmp_df)
```
